oss-internship-2020/libtiff/example/main_sandboxed.cc

```diff
@@ -0,0 +1,314 @@
+        << "Please run tests from build dir or send project dir as a "
+        << "parameter: ./sandboxed /absolute/path/to/project/dir .\n"
+        << "Falling back to using current working directory as root dir.\n";
+    project_path = cwd;
+  } else {
+    project_path = cwd.substr(0, find);
+  }
+
+  return sandbox2::file::JoinPath(project_path, "test", "images", filename);
+}
+
+absl::Status LibTIFFMain(const std::string& srcfile) {
+  // without addDir to sandbox. to add dir use
+  // sandbox(absolute_path_to_dir, srcfile) or
+  // sandbox(absolute_path_to_dir). file and dir should be exists.
+  // srcfile must also be absolute_path_to_file
+  TiffSapiSandbox sandbox(srcfile);
+
+  // initialize sapi vars after constructing TiffSapiSandbox
+  sapi::v::UShort h, v;
+  sapi::StatusOr<TIFF*> status_or_tif;
+  sapi::StatusOr<int> status_or_int;
+  sapi::StatusOr<tmsize_t> status_or_long;
+  absl::Status status;
+
+  status = sandbox.Init();
+
+  SAPI_RETURN_IF_ERROR(sandbox.Init());
+
+  TiffApi api(&sandbox);
+  sapi::v::ConstCStr srcfile_var(srcfile.c_str());
+  sapi::v::ConstCStr r_var("r");
+
+  SAPI_ASSIGN_OR_RETURN(
+      status_or_tif, api.TIFFOpen(srcfile_var.PtrBefore(), r_var.PtrBefore()));
+
+  sapi::v::RemotePtr tif(status_or_tif.value());
+  if (!tif.GetValue()) {
+    // tif is NULL
+    return absl::InternalError(absl::StrCat("Could not open ", srcfile));
+  }
+
+  SAPI_ASSIGN_OR_RETURN(auto return_value,
+                        api.TIFFGetField2(&tif, TIFFTAG_YCBCRSUBSAMPLING,
+                                          h.PtrBoth(), v.PtrBoth()));
+  if (return_value == 0 || h.GetValue() != 2 || v.GetValue() != 2) {
+    return absl::InternalError("Could not retrieve subsampling tag");
+  }
+
+  SAPI_ASSIGN_OR_RETURN(tsize_t sz, api.TIFFTileSize(&tif));
+  if (sz != 24576) {
+    return absl::InternalError(absl::StrCat("tiles are ", sz, " bytes\n"));
+  }
+
+  sapi::v::Array<uint8_t> buffer_(sz);
+  // Read a tile in decompressed form, but still YCbCr subsampled
+  SAPI_ASSIGN_OR_RETURN(
+      tsize_t new_sz,
+      api.TIFFReadEncodedTile(&tif, kRawTileNumber, buffer_.PtrBoth(), sz));
+  if (new_sz != sz) {
+    return absl::InternalError(absl::StrCat(
+        "Did not get expected result code from TIFFReadEncodedTile(): ",
+        status_or_long.value(), " instead of ", sz));
+  }
+
+  bool pixel_status = true;
+  if (status = CheckCluster(0, buffer_, kCluster0); !status.ok()) {
+    LOG(ERROR) << "CheckCluster failed:\n" << status.ToString();
+  }
+  pixel_status &= status.ok();
+
+  if (status = CheckCluster(64, buffer_, kCluster64); !status.ok()) {
+    LOG(ERROR) << "CheckCluster failed:\n" << status.ToString();
+  }
+  pixel_status &= status.ok();
+
+  if (status = CheckCluster(128, buffer_, kCluster128); !status.ok()) {
+    LOG(ERROR) << "CheckCluster failed:\n" << status.ToString();
+  }
+  pixel_status &= status.ok();
+
+  if (!pixel_status) {
+    return absl::InternalError("unexpected pixel_status value");
```

Comment:
This will only be the case if one or more of the cluster checks above failed, so maybe this message should state that instead (~One or more clusters failed the check)